core/src/test/java/io/grpc/internal/ManagedChannelImplTest.java

```diff
@@ -233,6 +239,98 @@ public void twoCallsAndGracefulShutdown() {
   }
 
   @Test
+  public void getNameResolver_invalidUriWithoutScheme() {
+    Factory nameResolverFactory =  new FakeNameResolverFactory(server, expectedUri, true);
+    thrown.expect(IllegalArgumentException.class);
+
+    ManagedChannelImpl.getNameResolver("[invalid", nameResolverFactory, Attributes.EMPTY);
+  }
+
+  @Test
+  public void getNameResolver_invalidUriWithScheme() {
+    Factory nameResolverFactory =  new FakeNameResolverFactory(server, expectedUri, true);
+    thrown.expect(IllegalArgumentException.class);
+
+    ManagedChannelImpl.getNameResolver("scheme://[invalid", nameResolverFactory, Attributes.EMPTY);
+  }
+
+  @Test
+  public void getNameResolver_validHost() {
+    Factory nameResolverFactory = new FakeNameResolverFactory(server, expectedUri, true);
+
+    NameResolver res = ManagedChannelImpl.getNameResolver(
+        target, nameResolverFactory, NAME_RESOLVER_PARAMS);
+
+    assertEquals(serviceName, res.getServiceAuthority());
+  }
+
+  @Test
+  public void getNameResolver_validHostWithoutSchema() throws URISyntaxException {
+    expectedUri = new URI("dns:///foo.googleapis.com:8080");
+    Factory nameResolverFactory = new NameResolver.Factory() {
+      @Override
+      public NameResolver newNameResolver(URI targetUri, Attributes params) {
+        if (targetUri.equals(expectedUri)) {
+          NameResolver resolver = mock(NameResolver.class);
+          when(resolver.getServiceAuthority()).thenReturn("127.0.0.1:8080");
```

Comment:
Though this doesn't affect the effectiveness of the test, "foo.googleapis.com:8080" would be what a real resolver would return.